src/cmd/services/m3coordinator/downsample/metrics_appender.go

```diff
@@ -370,6 +378,142 @@ func (a *metricsAppender) Finalize() {
 	a.pool.Put(a)
 }
 
+func (a *metricsAppender) newSamplesAppenders(
+	stagedMetadata metadata.StagedMetadata,
+) ([]samplesAppender, error) {
+	var (
+		pipelines []metadata.PipelineMetadata
+		appenders []samplesAppender
+	)
+	for _, pipeline := range stagedMetadata.Pipelines {
+		// For pipeline which have tags to augment we generate and send
+		// separate IDs. Other pipelines return the same.
+		pipeline := pipeline
+		if len(pipeline.Tags) == 0 && len(pipeline.GraphitePrefix) == 0 {
+			pipelines = append(pipelines, pipeline)
+			continue
+		}
+
+		tags := augmentTags(a.tags, pipeline.GraphitePrefix, pipeline.Tags, pipeline.AggregationID)
+
+		sm := stagedMetadata
+		sm.Pipelines = []metadata.PipelineMetadata{pipeline}
+
+		appender, err := a.newSamplesAppender(tags, sm, true)
+		if err != nil {
+			return []samplesAppender{}, err
+		}
+		appenders = append(appenders, appender)
+	}
+
+	if len(pipelines) == 0 {
+		return appenders, nil
+	}
+
+	sm := stagedMetadata
+	sm.Pipelines = pipelines
+
+	// NB(@shreyas): Passing in copyID as false assumes that this is the last
+	// call to newSamplesAppender. Any further calls will overwrite the ID.
+	// So adding any new calls to newSamplesAppender will need to change this.
+	appender, err := a.newSamplesAppender(a.tags, sm, false)
+	if err != nil {
+		return []samplesAppender{}, err
+	}
+	return append(appenders, appender), nil
+}
+
+func (a *metricsAppender) newSamplesAppender(
+	tags *tags,
+	sm metadata.StagedMetadata,
+	copyID bool,
+) (samplesAppender, error) {
+	a.tagEncoder.Reset()
+	if err := a.tagEncoder.Encode(tags); err != nil {
+		return samplesAppender{}, err
+	}
+	data, ok := a.tagEncoder.Data()
+	if !ok {
+		return samplesAppender{}, fmt.Errorf("unable to encode tags: names=%v, values=%v", tags.names, tags.values)
+	}
+	// NB (@shreyas): if this is the last use of the tagEncoder we do not need
+	// this copy. The caller tells us what to do here.
+	var id []byte
+	if copyID {
```

Comment:
As discussed, perhaps we can use two slices "cachedTagEncoders" and "inuseTagEncoders" to avoid in the regular case needing to alloc here for "extra" IDs?